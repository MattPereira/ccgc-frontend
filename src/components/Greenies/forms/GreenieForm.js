import React, { useState } from "react";
import CcgcApi from "../../../api/api";
import { useNavigate, useParams } from "react-router-dom";
import { Card, Form, Alert, Container, Row, Col } from "react-bootstrap";

import { Button, Typography } from "@mui/material";

/** Form to create a new greenie
 *
 *
 * Displays new greenie form and handles changes to local form state.
 * Submission of form calls the API to save the greenie.
 *
 * Redirects to CourseDetails page upon form submission.
 *
 * Routed as /greenies/:date/new
 * Routes -> NewGreenie -> NewGreenieForm
 */

const GreenieForm = ({ par3HoleNums, usernames, greenie }) => {
  let navigate = useNavigate();
  const { date } = useParams();

  /// HOW TO SEND AND SET THE ROUND ID AS SELECT INPUT OPTIONS CHANGE?
  /// NEED TO USE NAMES FOR THE SELECT INPUT OPTIONS USER SEES BUT KEEP ROUND ID IN STATE

  console.log(usernames);

  //dynamically set initial state of formData based on whether creating or updating
  //a greenie by looking to see if greenie is passed in as a prop
  const [formData, setFormData] = useState({
    roundId: greenie ? greenie.roundId : usernames[0][0],
    holeNumber: greenie ? greenie.holeNumber : par3HoleNums[0],
    feet: greenie ? greenie.feet : "",
    inches: greenie ? greenie.inches : "",
  });

  const [formErrors, setFormErrors] = useState([]);

  console.debug(
    "NewGreenieForm",
    "formData=",
    formData,
    "formErrors=",
    formErrors
  );
  //update state of formData onChange of any form input field
  const handleChange = (e) => {
    const { name, value } = e.target;

    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
    setFormErrors([]);
  };

  /**on form submission:
   * -attempt save to backend & report any errors
   * -if successful
   *  -clear previous error messages and password
   *  - show update-confirmed alert
   *  - set current user info throughout the site
   */

  const handleSubmit = async (e) => {
    e.preventDefault();

    let newGreenieData = {
      roundId: +formData.roundId,
      holeNumber: +formData.holeNumber,
      feet: +formData.feet,
      inches: +formData.inches,
    };

    let updateGreenieData = {
      feet: +formData.feet,
      inches: +formData.inches,
    };

    try {
      if (greenie) {
        await CcgcApi.updateGreenie(greenie.id, updateGreenieData);
      } else {
        await CcgcApi.createGreenie(newGreenieData);
      }
    } catch (errors) {
      debugger;
      setFormErrors(errors);
      return;
    }

    //navigate to the tournament detail page associated with the new greenie
    if (greenie) {
      navigate(`/greenies/${greenie.id}`);
    } else {
      navigate(`/tournaments/${date}`);
    }
  };

  return (
    <Container className="my-5">
      <div className="row justify-content-center">
        <div className="col-sm-10 col-md-8 col-lg-6">
          <Typography variant="h1" gutterBottom>
            {greenie ? "Edit" : "Add"} Greenie{" "}
          </Typography>
          <Card className="p-4">
            <Card.Body>
              <Form onSubmit={handleSubmit}>
                <Row className="mb-3 align-items-center">
                  <Col xs={2}>
                    <Form.Label htmlFor="roundId" className="mb-0">
                      <b>Name</b>
                    </Form.Label>
                  </Col>
                  <Col xs={10}>
                    {greenie ? (
                      <Form.Control
                        value={greenie.firstName + " " + greenie.lastName}
                        className="text-center"
                        readOnly
                      ></Form.Control>
                    ) : (
                      <Form.Select
                        name="roundId"
                        id="roundId"
                        value={formData.roundId}
                        onChange={handleChange}
                      >
                        {usernames.map((user) => (
                          <option key={user[0]} value={user[0]}>
                            {user[1]
                              .split("-")
                              .map((n) => n[0].toUpperCase() + n.slice(1))
                              .join(" ")}
                          </option>
                        ))}
                      </Form.Select>
                    )}
                  </Col>
                </Row>

                {greenie ? (
                  <Row className="mb-3 align-items-center">
                    <Col xs={2}>
                      <Form.Label htmlFor="roundId" className="mb-0">
                        <b>Date</b>
                      </Form.Label>
                    </Col>
                    <Col xs={10}>
                      <Form.Control
                        className="text-center"
                        value={new Date(
                          greenie.tournamentDate
                        ).toLocaleDateString("en-US", {
                          month: "long",
                          day: "numeric",
                          year: "numeric",
                        })}
                        readOnly
                      ></Form.Control>
                    </Col>
                  </Row>
                ) : null}

                <Row className="mb-3 align-items-center">
                  <Col xs={2}>
                    <Form.Label htmlFor="holeNumber" className="mb-0">
                      <b>Hole</b>
                    </Form.Label>
                  </Col>
                  <Col xs={10}>
                    {greenie ? (
                      <Form.Control
                        id="holeNumber"
                        name="holeNumber"
                        type="number"
                        value={greenie.holeNumber}
                        required
                        readOnly
                      />
                    ) : (
                      <Form.Select
                        className="form-control"
                        id="holeNumber"
                        name="holeNumber"
                        type="select"
                        onChange={handleChange}
                        value={formData.holeNumber}
                        required
                      >
                        {par3HoleNums.map((num) => (
                          <option key={num} value={num}>
                            {num}
                          </option>
                        ))}
                      </Form.Select>
                    )}
                  </Col>
                </Row>

                <Row className="mb-3 align-items-center">
                  <Col xs={2}>
                    <Form.Label htmlFor="feet" className="mb-0">
                      <b>Feet</b>
                    </Form.Label>
                  </Col>
                  <Col xs={10}>
                    <Form.Control
                      className="form-control"
                      id="feet"
                      name="feet"
                      type="number"
                      min="0"
                      onChange={handleChange}
                      value={formData.feet}
                      required
                    />
                  </Col>
                </Row>

                <Row className="mb-3 align-items-center">
                  <Col xs={2}>
                    <Form.Label htmlFor="inches" className="mb-0">
                      <b>Inches</b>
                    </Form.Label>
                  </Col>
                  <Col xs={10}>
                    <Form.Control
                      className="form-control"
                      id="inches"
                      name="inches"
                      type="number"
                      min="0"
                      max="11"
                      onChange={handleChange}
                      value={formData.inches}
                      required
                    />
                  </Col>
                </Row>

                {formErrors.length
                  ? formErrors.map((err) => (
                      <Alert key={err} variant="danger">
                        {err}
                      </Alert>
                    ))
                  : null}

                <div className="text-end">
                  <Button variant="contained" type="submit">
                    Submit
                  </Button>
                </div>
              </Form>
            </Card.Body>
          </Card>
        </div>
      </div>
    </Container>
  );
};

export default GreenieForm;
