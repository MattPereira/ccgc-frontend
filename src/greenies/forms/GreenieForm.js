import React, { useState } from "react";
import CcgcApi from "../../api/api";
import { useNavigate, useParams } from "react-router-dom";
import { Card, Form, Alert } from "react-bootstrap";

/** Form to create a new greenie
 *
 *
 * Displays new greenie form and handles changes to local form state.
 * Submission of form calls the API to save the greenie.
 *
 * Redirects to CourseDetails page upon form submission.
 *
 * Routed as /greenies/:date/new
 * Routes -> NewGreenie -> NewGreenieForm
 */

const GreenieForm = ({ par3HoleNums, playerName, greenie }) => {
  let navigate = useNavigate();
  const { roundId } = useParams();

  //dynamically set initial state of formData based on whether creating or updating
  //a greenie by looking to see if greenie is passed in as a prop
  const [formData, setFormData] = useState({
    roundId: greenie ? greenie.roundId : roundId,
    holeNumber: greenie ? greenie.holeNumber : par3HoleNums[0],
    feet: greenie ? greenie.feet : "",
    inches: greenie ? greenie.inches : "",
  });

  const [formErrors, setFormErrors] = useState([]);

  console.debug(
    "NewGreenieForm",
    "formData=",
    formData,
    "formErrors=",
    formErrors
  );
  //update state of formData onChange of any form input field
  const handleChange = (e) => {
    const { name, value } = e.target;

    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
    setFormErrors([]);
  };

  /**on form submission:
   * -attempt save to backend & report any errors
   * -if successful
   *  -clear previous error messages and password
   *  - show update-confirmed alert
   *  - set current user info throughout the site
   */

  const handleSubmit = async (e) => {
    e.preventDefault();

    let newGreenieData = {
      roundId: formData.roundId,
      holeNumber: formData.holeNumber,
      feet: formData.feet,
      inches: formData.inches,
    };

    let updateGreenieData = {
      feet: formData.feet,
      inches: formData.inches,
    };

    try {
      if (greenie) {
        await CcgcApi.updateGreenie(greenie.id, updateGreenieData);
      } else {
        await CcgcApi.createGreenie(newGreenieData);
      }
    } catch (errors) {
      debugger;
      setFormErrors(errors);
      return;
    }

    //navigate to the tournament detail page associated with the new greenie
    if (greenie) {
      navigate(`/greenies/${greenie.id}`);
    } else {
      navigate(`/rounds/${roundId}`);
    }
  };

  return (
    <div className="row justify-content-center">
      <div className="col-sm-10 col-md-8 col-lg-6">
        <Card>
          <Card.Title className="display-4 text-center bg-success py-2 text-white">
            {greenie ? "Edit" : "New"} Greenie
          </Card.Title>
          <Card.Body>
            <Form onSubmit={handleSubmit}>
              <Form.Group className="row align-items-center justify-content-center mb-3">
                <div className="col-3 col-lg-auto">
                  <Form.Label htmlFor="roundId" className="mb-0">
                    Name
                  </Form.Label>
                </div>
                <div className="col-9">
                  {greenie ? (
                    <Form.Control
                      value={greenie.firstName + " " + greenie.lastName}
                      className="text-center"
                      readOnly
                    ></Form.Control>
                  ) : (
                    <Form.Control value={playerName} readOnly></Form.Control>
                  )}
                </div>
              </Form.Group>

              {greenie ? (
                <Form.Group className="row align-items-center justify-content-center mb-3">
                  <div className="col-3 col-lg-auto">
                    <Form.Label htmlFor="roundId" className="mb-0">
                      Date
                    </Form.Label>
                  </div>
                  <div className="col-9">
                    <Form.Control
                      className="text-center"
                      value={new Date(
                        greenie.tournamentDate
                      ).toLocaleDateString()}
                      readOnly
                    ></Form.Control>
                  </div>
                </Form.Group>
              ) : null}

              <div className="">
                <Form.Group className="row justify-content-center align-items-center mb-3">
                  <div className="col-3 col-lg-auto">
                    <Form.Label htmlFor="holeNumber" className="mb-0">
                      Hole #
                    </Form.Label>
                  </div>
                  <div className="col-9">
                    {greenie ? (
                      <Form.Control
                        id="holeNumber"
                        name="holeNumber"
                        type="number"
                        value={greenie.holeNumber}
                        required
                        readOnly
                      />
                    ) : (
                      <Form.Select
                        className="form-control"
                        id="holeNumber"
                        name="holeNumber"
                        type="select"
                        onChange={handleChange}
                        value={formData.holeNumber}
                        required
                      >
                        {par3HoleNums.map((num) => (
                          <option key={num} value={num}>
                            {num}
                          </option>
                        ))}
                      </Form.Select>
                    )}
                  </div>
                </Form.Group>
              </div>
              <div className="">
                <Form.Group className="row justify-content-center align-items-center mb-3">
                  <div className="col-3 col-lg-auto">
                    <Form.Label htmlFor="feet" className="mb-0">
                      Feet
                    </Form.Label>
                  </div>
                  <div className="col-9">
                    <input
                      className="form-control"
                      id="feet"
                      name="feet"
                      type="number"
                      min="0"
                      onChange={handleChange}
                      value={formData.feet}
                      required
                    ></input>
                  </div>
                </Form.Group>
              </div>
              <div className="">
                <Form.Group className="row justify-content-center align-items-center mb-3">
                  <div className="col-3 col-lg-auto">
                    <Form.Label htmlFor="inches" className="mb-0">
                      Inches
                    </Form.Label>
                  </div>
                  <div className="col-9">
                    <input
                      className="form-control"
                      id="inches"
                      name="inches"
                      type="number"
                      min="0"
                      max="11"
                      onChange={handleChange}
                      value={formData.inches}
                      required
                    ></input>
                  </div>
                </Form.Group>
              </div>

              {formErrors.length
                ? formErrors.map((err) => (
                    <Alert key={err} variant="danger">
                      {err}
                    </Alert>
                  ))
                : null}

              <div className="row justify-content-end">
                <div className="col-auto">
                  <button className="btn btn-primary btn-block px-4">
                    Submit
                  </button>
                </div>
              </div>
            </Form>
          </Card.Body>
        </Card>
      </div>
    </div>
  );
};

export default GreenieForm;
