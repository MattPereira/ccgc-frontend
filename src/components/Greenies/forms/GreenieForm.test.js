import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import GreenieForm from "./GreenieForm";

const holeNums = [3, 7, 12, 16];

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <GreenieForm par3HoleNums={holeNums} />
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <GreenieForm par3HoleNums={holeNums} />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("displays form inputs and login button", function () {
  render(
    <MemoryRouter initialEntries={["/login"]}>
      <GreenieForm par3HoleNums={holeNums} />
    </MemoryRouter>
  );

  const holeNumberInput = screen.getByLabelText("Hole #");
  expect(holeNumberInput).toBeInTheDocument();

  const feetInput = screen.getByLabelText("Feet");
  expect(feetInput).toBeInTheDocument();

  const inchesInput = screen.getByLabelText("Inches");
  expect(inchesInput).toBeInTheDocument();

  //only one button in the component
  const submitBtn = screen.getByRole("button");
  expect(submitBtn).toBeInTheDocument();
});
