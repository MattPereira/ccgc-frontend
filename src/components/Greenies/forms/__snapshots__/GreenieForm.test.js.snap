// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row justify-content-center"
  >
    <div
      class="col-sm-10 col-md-8 col-lg-6"
    >
      <div
        class="card"
      >
        <div
          class="display-4 text-center bg-success py-2 text-white card-title h5"
        >
          New Greenie
        </div>
        <div
          class="card-body"
        >
          <form
            class=""
          >
            <div
              class="row align-items-center justify-content-center mb-3"
            >
              <div
                class="col-3 col-lg-auto"
              >
                <label
                  class="mb-0 form-label"
                  for="roundId"
                >
                  Name
                </label>
              </div>
              <div
                class="col-9"
              >
                <input
                  class="form-control"
                  readonly=""
                  value=""
                />
              </div>
            </div>
            <div
              class=""
            >
              <div
                class="row justify-content-center align-items-center mb-3"
              >
                <div
                  class="col-3 col-lg-auto"
                >
                  <label
                    class="mb-0 form-label"
                    for="holeNumber"
                  >
                    Hole #
                  </label>
                </div>
                <div
                  class="col-9"
                >
                  <select
                    class="form-control form-select"
                    id="holeNumber"
                    name="holeNumber"
                    required=""
                    type="select"
                  >
                    <option
                      value="3"
                    >
                      3
                    </option>
                    <option
                      value="7"
                    >
                      7
                    </option>
                    <option
                      value="12"
                    >
                      12
                    </option>
                    <option
                      value="16"
                    >
                      16
                    </option>
                  </select>
                </div>
              </div>
            </div>
            <div
              class=""
            >
              <div
                class="row justify-content-center align-items-center mb-3"
              >
                <div
                  class="col-3 col-lg-auto"
                >
                  <label
                    class="mb-0 form-label"
                    for="feet"
                  >
                    Feet
                  </label>
                </div>
                <div
                  class="col-9"
                >
                  <input
                    class="form-control"
                    id="feet"
                    min="0"
                    name="feet"
                    required=""
                    type="number"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class=""
            >
              <div
                class="row justify-content-center align-items-center mb-3"
              >
                <div
                  class="col-3 col-lg-auto"
                >
                  <label
                    class="mb-0 form-label"
                    for="inches"
                  >
                    Inches
                  </label>
                </div>
                <div
                  class="col-9"
                >
                  <input
                    class="form-control"
                    id="inches"
                    max="11"
                    min="0"
                    name="inches"
                    required=""
                    type="number"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="row justify-content-end"
            >
              <div
                class="col-auto"
              >
                <button
                  class="btn btn-primary btn-block px-4"
                >
                  Submit
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
